src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java

```diff
@@ -105,21 +108,46 @@ private void validateDirectPrerequisiteVisibility(
     if (!toolCheckAtDefinition
         || !attribute.isImplicit()
         || attribute.getName().equals(RuleClass.CONFIG_SETTING_DEPS_ATTRIBUTE)
-        || rule.getRuleClassObject().getRuleDefinitionEnvironmentLabel() == null) {
+        || !forStarlarkRuleOrAspect(context)) {
       // Default check: The attribute must be visible from the target.
       if (!context.isVisible(prerequisite.getConfiguredTarget())) {
         handleVisibilityConflict(context, prerequisite, rule.getLabel());
       }
     } else {
-      // For implicit attributes, check if the prerequisite is visible from the location of the
-      // rule definition
-      Label implicitDefinition = rule.getRuleClassObject().getRuleDefinitionEnvironmentLabel();
+      // For implicit attributes of Starlark rules or aspects, check if the prerequisite is visible
+      // from the location of the definition that declares the attribute. Only perform this check
+      // for the current aspect.
+      Label implicitDefinition;
+      Aspect mainAspect = context.getMainAspect();
+      if (mainAspect != null) {
+        if (!mainAspect.getDefinition().getAttributes().containsKey(attrName)) {
+          // Only verify visibility of implicit dependencies of the current aspect. Implicit
+          // dependencies of other aspects as well as the rule itself are checked when they are
+          // evaluated.
+          return;
```

Comment:
Readability: The `return` here breaks the encapsulation of the else block. You can't add any code after the block and assume it's correct.

Possible solution is to invert `if (!mainAspect.getDefinition().getAttributes().containsKey(attrName)) {`
and move the block `if (!RuleContext.isVisible...` inside after setting `implicitDefinition`.

Reply:
I wasn't entirely sure what you meant with "move the block ... inside". I left it outside without duplicating it, please let me know what you think.